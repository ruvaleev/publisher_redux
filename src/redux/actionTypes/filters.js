export const filtersActions = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_UNCOMPLETED: 'SHOW_UNCOMPLETED',
  SHOW_COMPLETED: 'SHOW_COMPLETED'
}
