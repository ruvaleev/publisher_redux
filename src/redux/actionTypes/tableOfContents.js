export const tableOfContentsActions = {
  TOGGLE_EDITABLE: 'TOGGLE_EDITABLE'
}
