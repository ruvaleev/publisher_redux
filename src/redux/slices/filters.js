import { createSlice } from '@reduxjs/toolkit';

const initialState = { value: 'SHOW_ALL' };

const filtersSlice = createSlice({
  name: 'filters',
  initialState,
  reducers: {
    showAll(state, action) {
      return { value: 'SHOW_ALL' }
    },
    showUncompleted(state, action) {
      return { value: 'SHOW_UNCOMPLETED' }
    },
    showCompleted(state, action) {
      return { value: 'SHOW_COMPLETED' }
    }
  }
})

export const { showAll, showUncompleted, showCompleted } = filtersSlice.actions;
export default filtersSlice.reducer;
