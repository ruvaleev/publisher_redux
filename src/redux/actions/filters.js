import { filtersActions } from '../actionTypes/filters';

export const showAll = () => ({
  type: filtersActions.SHOW_ALL
})

export const showUncompleted = () => ({
  type: filtersActions.SHOW_UNCOMPLETED
})

export const showCompleted = () => ({
  type: filtersActions.SHOW_COMPLETED
})
