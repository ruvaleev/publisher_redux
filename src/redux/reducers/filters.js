import { filtersActions } from "../actionTypes/filters";

const initialState = filtersActions.SHOW_ALL;

export const filtersReducer = function (state = initialState, action) {
  switch (action.type) {
    case filtersActions.SHOW_ALL:
      return filtersActions.SHOW_ALL
    case filtersActions.SHOW_UNCOMPLETED:
      return filtersActions.SHOW_UNCOMPLETED
    case filtersActions.SHOW_COMPLETED:
      return filtersActions.SHOW_COMPLETED
    default :
      return state
  }
};
