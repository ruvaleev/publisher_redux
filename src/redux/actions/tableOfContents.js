import { tableOfContentsActions } from '../actionTypes/tableOfContents';

export const toggleEditable = (editable) => ({
  editable,
  type: tableOfContentsActions.TOGGLE_EDITABLE
})
