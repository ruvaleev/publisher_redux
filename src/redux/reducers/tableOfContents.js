import { tableOfContentsActions } from "../actionTypes/tableOfContents";

const initialState = { editable: false };

export const tableOfContentsReducer = function (state = initialState, action) {
  switch (action.type) {
    case tableOfContentsActions.TOGGLE_EDITABLE :
      return {
        ...state,
        editable: !state.editable
      };
    default:
      return state;
  }
};
