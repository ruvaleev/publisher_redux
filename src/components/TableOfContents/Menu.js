import React from 'react';
import { connect } from 'react-redux';

import { toggleEditable } from '../../redux/actions/tableOfContents';
import { showAll, showUncompleted, showCompleted } from '../../redux/actions/filters';

function Menu ({editable, filter, showAll, showUncompleted, showCompleted, toggleEditable, undo, redo}) {
  return (
    <div className='fixed -m-16 text-xl'>
      <button className='px-4' onClick={() => toggleEditable(editable)}>Edit</button>
      <FilterButton filter={filter} title='Show All' callback={showAll}/>
      <FilterButton filter={filter} title='Show Uncompleted' callback={showUncompleted}/>
      <FilterButton filter={filter} title='Show Completed' callback={showCompleted}/>
      <button className='px-4' onClick={() => undo()}>Undo</button>
      <button className='px-4' onClick={() => redo()}>Redo</button>
    </div>
  )
}

function FilterButton ({filter, title, callback}) {
  return (
    <button
      className={`px-4 rounded-lg border-gray-500 border-solid ${filter == title && 'border'}`}
      onClick={() => callback()}>
      {title}
    </button>
  )
}

const mapStateToProps = (state) => ({
  editable: state.tableOfContentsReducer.editable,
  filter: state.filtersReducer.title
})

const mapDispatchToProps = (dispatch) => ({
  toggleEditable: (editable) => dispatch(toggleEditable(editable)),
  showAll: () => dispatch(showAll()),
  showUncompleted: () => dispatch(showUncompleted()),
  showCompleted: () => dispatch(showCompleted()),
})

export default connect(mapStateToProps, mapDispatchToProps)(Menu);
