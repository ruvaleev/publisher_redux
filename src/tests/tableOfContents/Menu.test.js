import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import Menu from '../../components/TableOfContents/Menu';
import Store from '../shared/Store';
import { toggleEditable } from '../../redux/slices/tableOfContents';
import { showAll, showUncompleted, showCompleted } from '../../redux/slices/filters';
import { ActionCreators } from 'redux-undo';

const mockStore = configureStore([]);
const leftClick = { button: 0 };
let store;
let component;

function renderWithStore(store) {
  return render(
    <Provider store={store}>
      <Menu/>
    </Provider>
  );
}

function renderWithDefaultStore() {
  store = mockStore(Store());
  return renderWithStore(store);
}

describe('Menu', () => {
  it('renders correctly Menu', () => {
    component = renderWithDefaultStore();

    expect(component).toMatchSnapshot();
  });
});

describe('Menu editable button', () => {
  it('dispatches toggleEditable on button click', () => {
    store = mockStore(Store());
    store.dispatch = jest.fn();
    component = renderWithStore(store);
    const editButton = screen.queryByText('Edit');
    fireEvent.click(editButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      toggleEditable(false)
    );
  });

  it('dispatches toggleEditable on button click with proper current editable flag', () => {
    store = mockStore(Store({ tableOfContentsReducer: { editable: true } }));
    store.dispatch = jest.fn();
    component = renderWithStore(store);
    const editButton = screen.queryByText('Edit');
    fireEvent.click(editButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      toggleEditable(true)
    );
  });
});

describe('Menu filters buttons', () => {
  beforeEach(() => {
    store = mockStore(Store());
    store.dispatch = jest.fn();
    component = renderWithStore(store);
  })

  it('dispatches Show All button click', () => {
    const showAllButton = screen.queryByText('Show All');
    fireEvent.click(showAllButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(showAll());
  });

  it('dispatches Show Uncompleted button click', () => {
    const showUncompletedButton = screen.queryByText('Show Uncompleted');
    fireEvent.click(showUncompletedButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(showUncompleted());
  });

  it('dispatches Show Completed button click', () => {
    const showCompletedButton = screen.queryByText('Show Completed');
    fireEvent.click(showCompletedButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(showCompleted());
  });
});

describe('Menu undoable', () => {
  beforeEach(() => {
    store = mockStore(Store());
    store.dispatch = jest.fn();
    component = renderWithStore(store);
  })

  it('dispatches Undo button click', () => {
    const undoButton = screen.queryByText('Undo');
    fireEvent.click(undoButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(ActionCreators.undo());
  });

  it('dispatches Redo button click', () => {
    const redoButton = screen.queryByText('Redo');
    fireEvent.click(redoButton, leftClick);
 
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(ActionCreators.redo());
  });
});
