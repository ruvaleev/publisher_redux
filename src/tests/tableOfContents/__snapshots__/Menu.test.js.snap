// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly Menu 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="fixed -m-16 text-xl"
      >
        <button
          class="px-4"
        >
          Edit
        </button>
        <button
          class="px-4 rounded-lg border-gray-500 border-solid border"
        >
          Show All
        </button>
        <button
          class="px-4 rounded-lg border-gray-500 border-solid false"
        >
          Show Uncompleted
        </button>
        <button
          class="px-4 rounded-lg border-gray-500 border-solid false"
        >
          Show Completed
        </button>
        <button
          class="px-4"
        >
          Undo
        </button>
        <button
          class="px-4"
        >
          Redo
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="fixed -m-16 text-xl"
    >
      <button
        class="px-4"
      >
        Edit
      </button>
      <button
        class="px-4 rounded-lg border-gray-500 border-solid border"
      >
        Show All
      </button>
      <button
        class="px-4 rounded-lg border-gray-500 border-solid false"
      >
        Show Uncompleted
      </button>
      <button
        class="px-4 rounded-lg border-gray-500 border-solid false"
      >
        Show Completed
      </button>
      <button
        class="px-4"
      >
        Undo
      </button>
      <button
        class="px-4"
      >
        Redo
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
